### ISSUE TITLE
Failure to serialise volatile shared objects.

### ISSUE THREAD

**Issue by pchilds** [USCiLab/cereal#601]
If I change a shared object while a JSONOutputArchive is open it will give no data:
```
std::ofstream archiveFile("file.json");
{
		auto A = std::make_shared<MyObject>(initialiser);
		cereal::JSONOutputArchive archiver(archiveFile);
		archive(cereal::make_nvp("Root", A));
		A->DoSomethingToIt();
		archive(cereal::make_nvp("Root", A));
}
archiveFile.flush();
```

The result is:
```
{
    "Root": {
        "ptr_wrapper": {
            "id": 2147483649,
            "data": {
                "cereal_class_version": 0,
                myData
            }
        }
    },
    "Root": {
        "ptr_wrapper": {
            "id": 1
        }
    }
}
```

**Comment by pchilds** [comment]
I just checked, this also fails with portable binary. I believe the trouble derives from the RAII design methodology.

**Comment by uentity** [comment]
This is absolutely correct behavior of not serializing same pointer twice.
The only thing Cereal does is checking pointer address. It has no idea that you have changed the pointee.

**Comment by pchilds** [comment]
Correct behaviour according to which design paradigm?

**Comment by uentity** [comment]
Correct behaviour considering most of similar serialization frameworks like `boost::serialization`, etc.

You're asking a serialization framework to track changes in object addressed by shared pointer, am I right? But for doing so in constrain of dumping only unique objects would at least require to call some object's comparison operator to judge whether modification happened or not.
Now imagine, that object contains an array of 1M scalar values. You will have to iterate over all values to understand if object was modified.

In other words, tracking object modifications will require comparison (or hashing) operator defined and may have unpredictable complexity that can be infinite.

**Comment by uentity** [comment]
Also try to answer what must happen on deserialization stage if we earlier serialized multiple pointers to same object and have multiple different states of this same object in archive?

**Comment by pchilds** [comment]
I was using boost for serialisation before moving to cereal. I moved because boost binary archives were not portable. The code base otherwise worked fine. If I knew about this problem and 598 I would have never chosen cereal.

I am not asking cereal to track changes. I want it to serialise when I tell it to serialise. I want my deserialisation to match what I serialise. Boost did this for me. Cereal does not and there is no documentstion to indicate the many gotchas I am facing.

**Comment by erichkeane** [comment]
If you depend on that behavior, don't types/cereal/memory.hpp and write your own shared_ptr archiver.  Cereal is designed to do 'all at once' serialization, so things like this pointer-optimization for size purposes are permitted.  If you don't like that part, you don't have to use it.

**Comment by pchilds** [comment]
I think you should make it explicit in the documentation that cereal only supports all at once serialisation.
